File: inference_profiler/collectors/container.py
import os
from .base import BaseCollector


class ContainerCollector(BaseCollector):
    CGROUP_DIR = '/sys/fs/cgroup'

    @staticmethod
    def is_v2():
        return os.path.exists(os.path.join(ContainerCollector.CGROUP_DIR, "cgroup.controllers"))

    @staticmethod
    def collect():
        if not os.path.exists(ContainerCollector.CGROUP_DIR):
            return {}
        if ContainerCollector.is_v2():
            return ContainerCollector._collect_v2()
        else:
            return ContainerCollector._collect_v1()

    @staticmethod
    def _get_container_id():
        """Attempt to detect container ID from various sources."""
        # Try /proc/self/cgroup
        try:
            with open('/proc/self/cgroup', 'r') as f:
                for line in f:
                    parts = line.strip().split(':')
                    if len(parts) >= 3:
                        path = parts[2]
                        # Docker format: /docker/<container_id>
                        if '/docker/' in path:
                            return path.split('/docker/')[-1][:12]
                        # Kubernetes format: various patterns
                        if '/kubepods/' in path:
                            segments = path.split('/')
                            if segments:
                                return segments[-1][:12]
        except Exception:
            pass

        # Try hostname as fallback (often set to container ID)
        try:
            hostname = os.uname().nodename
            if len(hostname) == 12 and all(c in '0123456789abcdef' for c in hostname.lower()):
                return hostname
        except Exception:
            pass

        return "unavailable"

    @staticmethod
    def _collect_v1():
        cpu_path = os.path.join(ContainerCollector.CGROUP_DIR, "cpuacct", "cpuacct.usage")
        cpu_stat_path = os.path.join(ContainerCollector.CGROUP_DIR, "cpuacct", "cpuacct.stat")
        mem_path = os.path.join(ContainerCollector.CGROUP_DIR, "memory", "memory.usage_in_bytes")
        mem_max_path = os.path.join(ContainerCollector.CGROUP_DIR, "memory", "memory.max_usage_in_bytes")
        blkio_path = os.path.join(ContainerCollector.CGROUP_DIR, "blkio", "blkio.throttle.io_service_bytes")

        # CPU metrics
        cpu_usage, t_cpu = BaseCollector._read_int(cpu_path)

        # CPU user/kernel split from cpuacct.stat
        cpu_stat, t_cpu_stat = BaseCollector._parse_proc_kv(cpu_stat_path, separator=' ')
        cpu_user_jiffies = int(cpu_stat.get('user', 0))
        cpu_kernel_jiffies = int(cpu_stat.get('system', 0))

        # Memory metrics
        mem_usage, t_mem = BaseCollector._read_int(mem_path)
        mem_max, t_mem_max = BaseCollector._read_int(mem_max_path)

        # Block I/O metrics
        disk_read, disk_write, t_blkio = ContainerCollector._parse_blkio_v1(blkio_path)

        # Per-CPU stats
        per_cpu = ContainerCollector._get_per_cpu_v1()

        # Network stats (from container's network namespace)
        net_recv, net_sent, t_net = ContainerCollector._get_container_net_stats()

        result = {
            "cId": ContainerCollector._get_container_id(),
            "cCgroupVersion": 1,
            # Total CPU time in nanoseconds
            "cCpuTime": cpu_usage,
            "tcCpuTime": t_cpu,
            # CPU time split by mode (in centiseconds)
            "cCpuTimeUserMode": cpu_user_jiffies * BaseCollector.JIFFIES_PER_SECOND,
            "tcCpuTimeUserMode": t_cpu_stat,
            "cCpuTimeKernelMode": cpu_kernel_jiffies * BaseCollector.JIFFIES_PER_SECOND,
            "tcCpuTimeKernelMode": t_cpu_stat,
            # Number of processors
            "cNumProcessors": os.cpu_count() or 0,
            # Memory metrics
            "cMemoryUsed": mem_usage,
            "tcMemoryUsed": t_mem,
            "cMemoryMaxUsed": mem_max,
            "tcMemoryMaxUsed": t_mem_max,
            # Disk I/O
            "cDiskReadBytes": disk_read,
            "tcDiskReadBytes": t_blkio,
            "cDiskWriteBytes": disk_write,
            "tcDiskWriteBytes": t_blkio,
            # Network
            "cNetworkBytesRecvd": net_recv,
            "tcNetworkBytesRecvd": t_net,
            "cNetworkBytesSent": net_sent,
            "tcNetworkBytesSent": t_net,
        }

        # Add per-CPU stats
        result.update(per_cpu)

        return result

    @staticmethod
    def _collect_v2():
        cpu_stat_path = os.path.join(ContainerCollector.CGROUP_DIR, "cpu.stat")
        mem_path = os.path.join(ContainerCollector.CGROUP_DIR, "memory.current")
        mem_peak_path = os.path.join(ContainerCollector.CGROUP_DIR, "memory.peak")
        io_stat_path = os.path.join(ContainerCollector.CGROUP_DIR, "io.stat")

        # Memory metrics
        mem_usage, t_mem = BaseCollector._read_int(mem_path)
        mem_peak, t_mem_peak = BaseCollector._read_int(mem_peak_path)

        # Parse cpu.stat key-values
        cpu_stats, t_cpu = BaseCollector._parse_proc_kv(cpu_stat_path, separator=' ')
        cpu_usage_micros = int(cpu_stats.get('usage_usec', 0))
        cpu_user_micros = int(cpu_stats.get('user_usec', 0))
        cpu_system_micros = int(cpu_stats.get('system_usec', 0))

        # I/O stats
        disk_read, disk_write, t_io = ContainerCollector._parse_io_stat_v2(io_stat_path)

        # Network stats
        net_recv, net_sent, t_net = ContainerCollector._get_container_net_stats()

        return {
            "cId": ContainerCollector._get_container_id(),
            "cCgroupVersion": 2,
            # Total CPU time in nanoseconds (converted from microseconds)
            "cCpuTime": cpu_usage_micros * 1000,
            "tcCpuTime": t_cpu,
            # CPU time split by mode (in centiseconds, converted from microseconds)
            "cCpuTimeUserMode": cpu_user_micros // 10000,  # usec to centisec
            "tcCpuTimeUserMode": t_cpu,
            "cCpuTimeKernelMode": cpu_system_micros // 10000,
            "tcCpuTimeKernelMode": t_cpu,
            # Number of processors
            "cNumProcessors": os.cpu_count() or 0,
            # Memory metrics
            "cMemoryUsed": mem_usage,
            "tcMemoryUsed": t_mem,
            "cMemoryMaxUsed": mem_peak,
            "tcMemoryMaxUsed": t_mem_peak,
            # Disk I/O
            "cDiskReadBytes": disk_read,
            "tcDiskReadBytes": t_io,
            "cDiskWriteBytes": disk_write,
            "tcDiskWriteBytes": t_io,
            # Network
            "cNetworkBytesRecvd": net_recv,
            "tcNetworkBytesRecvd": t_net,
            "cNetworkBytesSent": net_sent,
            "tcNetworkBytesSent": t_net,
        }

    @staticmethod
    def _get_per_cpu_v1():
        """Get per-CPU usage for cgroup v1."""
        result = {}
        percpu_path = os.path.join(ContainerCollector.CGROUP_DIR, "cpuacct", "cpuacct.usage_percpu")
        content, timestamp = BaseCollector._probe_file(percpu_path)

        if content:
            values = content.split()
            for i, val in enumerate(values):
                try:
                    result[f"cCpu{i}Time"] = int(val)
                    result[f"tcCpu{i}Time"] = timestamp
                except ValueError:
                    continue

        return result

    @staticmethod
    def _parse_blkio_v1(path):
        """Parse blkio.throttle.io_service_bytes for v1."""
        read_bytes = 0
        write_bytes = 0

        lines, timestamp = BaseCollector._get_file_lines(path)
        for line in lines:
            parts = line.split()
            if len(parts) >= 3:
                try:
                    op = parts[1].lower()
                    value = int(parts[2])
                    if op == 'read':
                        read_bytes += value
                    elif op == 'write':
                        write_bytes += value
                except (ValueError, IndexError):
                    continue

        return read_bytes, write_bytes, timestamp

    @staticmethod
    def _parse_io_stat_v2(path):
        """Parse io.stat for cgroup v2."""
        read_bytes = 0
        write_bytes = 0

        lines, timestamp = BaseCollector._get_file_lines(path)
        for line in lines:
            # Format: "major:minor rbytes=X wbytes=Y rios=Z wios=W ..."
            parts = line.split()
            for part in parts:
                if part.startswith('rbytes='):
                    try:
                        read_bytes += int(part.split('=')[1])
                    except (ValueError, IndexError):
                        pass
                elif part.startswith('wbytes='):
                    try:
                        write_bytes += int(part.split('=')[1])
                    except (ValueError, IndexError):
                        pass

        return read_bytes, write_bytes, timestamp

    @staticmethod
    def _get_container_net_stats():
        """Get network stats from container's perspective."""
        recv = 0
        sent = 0

        lines, timestamp = BaseCollector._get_file_lines('/proc/net/dev')

        for line in lines[2:]:  # Skip headers
            if ':' in line:
                try:
                    iface, data = line.split(':', 1)
                    iface = iface.strip()

                    # Skip loopback
                    if iface == 'lo':
                        continue

                    parts = data.split()
                    if len(parts) >= 9:
                        recv += int(parts[0])
                        sent += int(parts[8])
                except (ValueError, IndexError):
                    continue

        return recv, sent, timestamp

File: src/container.go
package src

import (
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// ContainerMetrics contains container-level measurements from cgroups.
type ContainerMetrics struct {
	// Identification
	ID            string `json:"cId"`            // Container ID (Docker/K8s short ID)
	CgroupVersion int    `json:"cCgroupVersion"` // 1 or 2

	// CPU (time in nanoseconds for total, centiseconds for user/kernel)
	CPUTime       exporter.Timed[int64] `json:"cCpuTime"`           // Total CPU nanoseconds
	CPUUserMode   exporter.Timed[int64] `json:"cCpuTimeUserMode"`   // User mode centiseconds
	CPUKernelMode exporter.Timed[int64] `json:"cCpuTimeKernelMode"` // Kernel mode centiseconds
	NumProcessors int                   `json:"cNumProcessors"`     // Available CPUs

	// Per-CPU times (cgroup v1 only)
	PerCPU map[string]exporter.Timed[int64] `json:"perCpu,omitempty"` // cCpu{i}Time in nanoseconds

	// Memory (bytes)
	MemoryUsed    exporter.Timed[int64] `json:"cMemoryUsed"`    // Current usage
	MemoryMaxUsed exporter.Timed[int64] `json:"cMemoryMaxUsed"` // Peak usage

	// Disk I/O (bytes)
	DiskReadBytes  exporter.Timed[int64] `json:"cDiskReadBytes"`
	DiskWriteBytes exporter.Timed[int64] `json:"cDiskWriteBytes"`

	// Network (bytes, from container namespace)
	NetBytesRecvd exporter.Timed[int64] `json:"cNetworkBytesRecvd"`
	NetBytesSent  exporter.Timed[int64] `json:"cNetworkBytesSent"`
}

const cgroupDir = "/sys/fs/cgroup"

// CollectContainer gathers container metrics from cgroups.
// Auto-detects cgroup v1 vs v2.
func CollectContainer() ContainerMetrics {
	if _, err := os.Stat(cgroupDir); err != nil {
		return ContainerMetrics{}
	}

	if isCgroupV2() {
		return collectCgroupV2()
	}
	return collectCgroupV1()
}

// isCgroupV2 checks for cgroup v2 unified hierarchy.
func isCgroupV2() bool {
	_, err := os.Stat(filepath.Join(cgroupDir, "cgroup.controllers"))
	return err == nil
}

// getContainerID attempts to detect container ID from /proc/self/cgroup or hostname.
func getContainerID() string {
	lines, _ := exporter.readLines("/proc/self/cgroup")
	for _, line := range lines {
		parts := strings.SplitN(line, ":", 3)
		if len(parts) < 3 {
			continue
		}
		path := parts[2]

		// Docker format: /docker/<container_id>
		if idx := strings.Index(path, "/docker/"); idx >= 0 {
			id := path[idx+8:]
			if len(id) > 12 {
				id = id[:12]
			}
			return id
		}

		// Kubernetes format: /kubepods/.../<container_id>
		if strings.Contains(path, "/kubepods/") {
			segments := strings.Split(path, "/")
			if len(segments) > 0 {
				id := segments[len(segments)-1]
				if len(id) > 12 {
					id = id[:12]
				}
				return id
			}
		}
	}

	// Fallback: hostname (often set to container ID)
	hostname, _ := os.Hostname()
	if len(hostname) == 12 && isHex(hostname) {
		return hostname
	}

	return "unavailable"
}

func isHex(s string) bool {
	for _, c := range strings.ToLower(s) {
		if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')) {
			return false
		}
	}
	return true
}

// collectCgroupV1 gathers metrics from cgroup v1 hierarchy.
func collectCgroupV1() ContainerMetrics {
	cpuUsage, cpuTS := exporter.readInt(filepath.Join(cgroupDir, "cpuacct", "cpuacct.usage"))

	cpuStat, cpuStatTS := exporter.parseKV(filepath.Join(cgroupDir, "cpuacct", "cpuacct.stat"), ' ')
	userJiffies, _ := strconv.ParseInt(cpuStat["user"], 10, 64)
	systemJiffies, _ := strconv.ParseInt(cpuStat["system"], 10, 64)

	memUsage, memTS := exporter.readInt(filepath.Join(cgroupDir, "memory", "memory.usage_in_bytes"))
	memMax, memMaxTS := exporter.readInt(filepath.Join(cgroupDir, "memory", "memory.max_usage_in_bytes"))

	diskRead, diskWrite, blkioTS := parseBlkioV1(filepath.Join(cgroupDir, "blkio", "blkio.throttle.io_service_bytes"))

	netRecv, netSent, netTS := getContainerNetStats()

	perCPU := getPerCPUV1()

	return ContainerMetrics{
		ID:             getContainerID(),
		CgroupVersion:  1,
		CPUTime:        exporter.TimedAt(cpuUsage, cpuTS),
		CPUUserMode:    exporter.TimedAt(userJiffies*jiffiesPerSec, cpuStatTS),
		CPUKernelMode:  exporter.TimedAt(systemJiffies*jiffiesPerSec, cpuStatTS),
		NumProcessors:  getNumCPU(),
		PerCPU:         perCPU,
		MemoryUsed:     exporter.TimedAt(memUsage, memTS),
		MemoryMaxUsed:  exporter.TimedAt(memMax, memMaxTS),
		DiskReadBytes:  exporter.TimedAt(diskRead, blkioTS),
		DiskWriteBytes: exporter.TimedAt(diskWrite, blkioTS),
		NetBytesRecvd:  exporter.TimedAt(netRecv, netTS),
		NetBytesSent:   exporter.TimedAt(netSent, netTS),
	}
}

// collectCgroupV2 gathers metrics from cgroup v2 unified hierarchy.
func collectCgroupV2() ContainerMetrics {
	cpuStat, cpuTS := exporter.parseKV(filepath.Join(cgroupDir, "cpu.stat"), ' ')
	usageUsec, _ := strconv.ParseInt(cpuStat["usage_usec"], 10, 64)
	userUsec, _ := strconv.ParseInt(cpuStat["user_usec"], 10, 64)
	systemUsec, _ := strconv.ParseInt(cpuStat["system_usec"], 10, 64)

	memUsage, memTS := exporter.readInt(filepath.Join(cgroupDir, "memory.current"))
	memPeak, memPeakTS := exporter.readInt(filepath.Join(cgroupDir, "memory.peak"))

	diskRead, diskWrite, ioTS := parseIOStatV2(filepath.Join(cgroupDir, "io.stat"))

	netRecv, netSent, netTS := getContainerNetStats()

	return ContainerMetrics{
		ID:             getContainerID(),
		CgroupVersion:  2,
		CPUTime:        exporter.TimedAt(usageUsec*1000, cpuTS), // usec -> ns
		CPUUserMode:    exporter.TimedAt(userUsec/10000, cpuTS), // usec -> cs
		CPUKernelMode:  exporter.TimedAt(systemUsec/10000, cpuTS),
		NumProcessors:  getNumCPU(),
		MemoryUsed:     exporter.TimedAt(memUsage, memTS),
		MemoryMaxUsed:  exporter.TimedAt(memPeak, memPeakTS),
		DiskReadBytes:  exporter.TimedAt(diskRead, ioTS),
		DiskWriteBytes: exporter.TimedAt(diskWrite, ioTS),
		NetBytesRecvd:  exporter.TimedAt(netRecv, netTS),
		NetBytesSent:   exporter.TimedAt(netSent, netTS),
	}
}

// getPerCPUV1 reads per-CPU usage from cgroup v1.
func getPerCPUV1() map[string]exporter.Timed[int64] {
	result := make(map[string]exporter.Timed[int64])
	content, ts := exporter.readFile(filepath.Join(cgroupDir, "cpuacct", "cpuacct.usage_percpu"))
	if content == "" {
		return result
	}

	fields := strings.Fields(content)
	for i, val := range fields {
		v, err := strconv.ParseInt(val, 10, 64)
		if err != nil {
			continue
		}
		key := "cCpu" + strconv.Itoa(i) + "Time"
		result[key] = exporter.TimedAt(v, ts)
	}
	return result
}

// parseBlkioV1 parses blkio.throttle.io_service_bytes for v1.
func parseBlkioV1(path string) (read, write, ts int64) {
	lines, ts := exporter.readLines(path)
	for _, line := range lines {
		fields := strings.Fields(line)
		if len(fields) < 3 {
			continue
		}
		op := strings.ToLower(fields[1])
		val, _ := strconv.ParseInt(fields[2], 10, 64)
		switch op {
		case "read":
			read += val
		case "write":
			write += val
		}
	}
	return read, write, ts
}

// parseIOStatV2 parses io.stat for cgroup v2.
func parseIOStatV2(path string) (read, write, ts int64) {
	lines, ts := exporter.readLines(path)
	for _, line := range lines {
		// Format: "major:minor rbytes=X wbytes=Y rios=Z wios=W ..."
		fields := strings.Fields(line)
		for _, f := range fields {
			if strings.HasPrefix(f, "rbytes=") {
				val, _ := strconv.ParseInt(f[7:], 10, 64)
				read += val
			} else if strings.HasPrefix(f, "wbytes=") {
				val, _ := strconv.ParseInt(f[7:], 10, 64)
				write += val
			}
		}
	}
	return read, write, ts
}

// getContainerNetStats reads /proc/net/dev from container's namespace.
func getContainerNetStats() (recv, sent, ts int64) {
	lines, ts := exporter.readLines("/proc/net/dev")

	// Skip headers
	for i := 2; i < len(lines); i++ {
		line := lines[i]
		colonIdx := strings.Index(line, ":")
		if colonIdx < 0 {
			continue
		}

		iface := strings.TrimSpace(line[:colonIdx])
		if iface == "lo" {
			continue
		}

		fields := strings.Fields(line[colonIdx+1:])
		if len(fields) < 9 {
			continue
		}

		r, _ := strconv.ParseInt(fields[0], 10, 64)
		s, _ := strconv.ParseInt(fields[8], 10, 64)
		recv += r
		sent += s
	}
	return recv, sent, ts
}
